### PR TITLE
Make input handling teehistorian friendly

### PR DESCRIPTION
**Pull request by Zwelf**
Makes input handling more reproducible during respawn. Still WIP, because I haven't tested this particular patch (only a different one with the same effect).

## Checklist

- [x] Tested the change ingame
- [ ] Provided screenshots if it is a visual change
- [x] Tested in combination with possibly related configuration options
  - [x] swap
  - [x] using timeout protection
  - [x] reconnecting
  - [x] respawning
  - [x] chat keeps movement
  - [x] pause keeps movement
- [x] Written a unit test if it works standalone, system.c especially
- [x] Considered possible null pointers and out of bounds array indexing
- [x] Changed no physics that affect existing maps
- [x] Tested the change with [ASan+UBSan or valgrind's memcheck](https://github.com/ddnet/ddnet/#using-addresssanitizer--undefinedbehavioursanitizer-or-valgrinds-memcheck) (optional)

### CODE REVIEW

File: src/game/client/prediction/entities/character.cpp [hunk 1]
```diff
@@ -527,7 +527,7 @@ void CCharacter::OnDirectInput(CNetObj_PlayerInput *pNewInput)
 	if(m_LatestInput.m_TargetX == 0 && m_LatestInput.m_TargetY == 0)
 		m_LatestInput.m_TargetY = -1;
 
-	if(m_NumInputs > 2 && Team() != TEAM_SPECTATORS)
+	if(m_NumInputs > 1 && Team() != TEAM_SPECTATORS)
```
Comment (heinrich5991) [inline src/game/client/prediction/entities/character.cpp:530]: What's the reason for this change?

File: src/game/server/gamecontext.cpp [hunk 2]
```diff
@@ -1133,7 +1133,7 @@ void CGameContext::OnClientPredictedEarlyInput(int ClientID, void *pInput)
 	if(!m_World.m_Paused)
 		m_apPlayers[ClientID]->OnPredictedEarlyInput((CNetObj_PlayerInput *)pInput);
 
-	if(m_TeeHistorianActive)
+	if(m_TeeHistorianActive && pInput != nullptr)
```
Comment (heinrich5991) [inline src/game/server/gamecontext.cpp:1136]: This means that in some ticks, no input is passed to teehistorian? What's the idea here?

File: src/game/client/prediction/entities/character.cpp (hunk 1 above)
Comment (Zwelf) [reply src/game/client/prediction/entities/character.cpp:530]: It's just something I noticed in the code. The weapon code (this is the only code using this variable) requires that the LatestPrevInput and PrevInput is there to calculate the number of weapon switches & fire button presses. I think it is a off by one error in the previous code (requiring three sent inputs instead of two) and wanted to fix it in the same chain. But it is not required to change, just nice to have for me.

But I'll also have to move the `m_NumInputs++` down into this function to count the inputs correctly.

File: src/game/server/gamecontext.cpp (hunk 2 above)
Comment (Zwelf) [reply src/game/server/gamecontext.cpp:1136]: Currently also on some ticks no input is passed to teehistorian, but I can change it to pass the last input stored in that variable. Sounds like a good idea.

File: src/game/client/prediction/entities/character.cpp (hunk 1 above)
Comment (Zwelf) [reply src/game/client/prediction/entities/character.cpp:530]: Moved the `m_NumInputs++` to the correct place.

File: src/game/server/gamecontext.cpp (hunk 2 above)
Comment (Zwelf) [reply src/game/server/gamecontext.cpp]: Added another fix I wanted to do: Use the NEW_INPUT for new players on an existing player slot. Haven't adjusted the test cases yet, but will follow.

**Comment by Zwelf** [comment]
Testing still missing, but I'm already happier with this version than the previous. I would like to bundle a change to record player swaps. Maybe I'll put all changes into one commit to make reverting easier (it's an input change after all, things can go wrong).

Changed input even more with https://github.com/ddnet/ddnet/pull/5032/commits/b64c2b53267963fb9a699aaa4537289c4a03a869.

**Comment by heinrich5991** [comment]
> Maybe I'll put all changes into one commit to make reverting easier (it's an input change after all, things can go wrong).

I don't think that is needed.

**Comment by Zwelf** [comment]
> * [x]  chat keeps movement until death
> 
> * [x]  pause keeps movement until death

After further though, movement should be kept even after death to always apply the last send input. So I'll investigate to change that.

**Comment by Zwelf** [comment]
* [x] Tested that movement is kept on respawn when chatting or pausing with last patch applied

**Comment by C0D3D3V** [comment]
> Tested that movement is kept on respawn when chatting or pausing with last patch applied

have you pushed this already? In your current version it is not fixed

**Comment by Zwelf** [comment]
I'm pretty sure that it I've done this in https://github.com/ddnet/ddnet/pull/5032/commits/971f2cfaa934fb3f1448146d836fbb1eb4338c96, but let me validate or fix it.

**Comment by C0D3D3V** [comment]
but we are talking about the same bug: https://github.com/ddnet/ddnet/issues/4653 ?

**Comment by Zwelf** [comment]
Ah true, this change is about different behavior. This PR is about keeping movement after respawn. (Does not change the behavior when leaving chat).

**Comment by Zwelf** [comment]
Example of what https://github.com/ddnet/ddnet/commit/971f2cfaa934fb3f1448146d836fbb1eb4338c96 change.
1. On the map Careless at the spawn
2. hold d and enter chat (or pause)
3. The tee walks into death tiles
4. after respawn observe that the tee keeps moving

Previously:

1. On the map Careless at the spawn
2. hold d and enter chat (or pause)
3. The tee walks into death tiles
4. after respawn observe that the tee stopped moving
5. when leaving the pause mode (not chat), the tee starts moving again

**Comment by Zwelf** [comment]
Resolved conflict with https://github.com/ddnet/ddnet/commit/f5ff2275617eb2e2bcd947bc4c5ca846e2b642ef#diff-4b7e0de54578478dbb720a146e73785264c55592e8ba622a241e7eb0af5e2975, which I also included in https://github.com/ddnet/ddnet/pull/5032/commits/416af01e0a82cc1d6cb8f9e8da266e60c7d63fde

+ added commit simplifying an expression made possible by the change.

File: src/game/server/entities/character.cpp [hunk 3]
```diff
@@ -19,14 +19,15 @@
...
 // Character, "physical" player's part
-CCharacter::CCharacter(CGameWorld *pWorld) :
+CCharacter::CCharacter(CGameWorld *pWorld, CNetObj_PlayerInput LastInput) :
 	CEntity(pWorld, CGameWorld::ENTTYPE_CHARACTER, vec2(0, 0), ms_PhysSize)
 {
 	m_Health = 0;
 	m_Armor = 0;
 	m_StrongWeakID = 0;
 
 	// never intilize both to zero
```
Comment (def-) [inline src/game/server/entities/character.cpp:29]: initialize*, and should be one line lower I guess

**Comment by def-** [comment]
Otherwise fine with me. I hope no physics changes will be caused.

**Comment by Zwelf** [comment]
This shouldn't change physics, because it only makes inputting more consistent during spawn. Previously it depended on the timing of the packets when the latest input was applied during spawn. But the new consistent variant was already possibility.

I've addressed the review in a new commit. I can rebase it into another commit if necessary, but I didn't want to invalidate the review :)

**Comment by def-** [comment]
bors r+
Thanks for the cleanup!

**Comment by fokkonaut** [comment]
It applies old input when none is sent? Sounds bad for inactive dummys?

**Comment by Zwelf** [comment]
Can you elaborate?
* When no input was sent at all, no input gets applied. (this only happens immediately after joining until the first input)
* The activity check (displaying the Zzz and closing the eyes) is done via input change. So this isn't affected.

**Comment by fokkonaut** [comment]
> When no input was sent at all, no input gets applied. (this only happens immediately after joining until the first input)
This is also the case for when dummy hammer is active, it only sends inputs on hammer

**Comment by C0D3D3V** [comment]
Deep fly + Dummy copy seems to be broken with this change

**Comment by Zwelf** [comment]
I can confirm :/ Should be reverted

**Comment by C0D3D3V** [comment]
Or you fix the client/server side so that it works again ^^

**Comment by Zwelf** [comment]
> This is also the case for when dummy hammer is active, it only sends inputs on hammer

Still trying to understand why that matters... Freeze works as expected (dummy hammering back). Deep freeze not.

**Comment by Zwelf** [comment]
Hm, sorry again. Somehow on my local test server on `Just deep fly` was set to `sv_deepfly 0`. setting `sv_deepfly 1` again let the deepfly work again (probably old ddnet-maps repository). So I think it works as I previously thought and doesn't need to be reverted.

 @C0D3D3V I'll test your config now too. Sorry for the back and forth.

**Comment by Zwelf** [comment]
> bind mouse1 "+fire; +toggle cl_dummy_hammer 1 0"
> and in addition turn dummy copy on

As far as I can tell this bind is also working on master with this PR merged.

**Comment by C0D3D3V** [comment]
nope I can make a video, one moment

**Comment by C0D3D3V** [comment]
Ah lol I was stupid, I compared with 15.9.1 :D and there is the combination broken ^^ not in master :D I mixed up on what screen I had open which version

**Comment by C0D3D3V** [comment]
this https://youtu.be/KKWD3CibF2w
maybe we ask a good dummy player to test most usecases of dummy inputs

**Comment by C0D3D3V** [comment]
Mh no I think it is broken. Because if you start deepfly (with undepp dummy), and then turn on copy, the dummy switches to gun, instead of staying with hammer.

**Comment by C0D3D3V** [comment]
So I tested all dummy controls, and murpi told me the correct bind for the copy fly, and everything seems to work as before (I compared with 15.9.1 and 16.0.3).

**Comment by Zwelf** [comment]
Thanks for investigating :) I'm glad that it turned out this way.

**Comment by def-** [comment]
@Zwelf Spooky reported that this changes physics on the map Fireball. @C0D3D3V confirmed it. Anything we want to do about that or do we just accept the new physics?

**Comment by Chairn** [comment]
I can confirm that this PR makes the map `Bootcamp #1` quite hard to finish. Given that it also slightly changes physics, maybe we should revert it?

@Zwelf i haven't tried it yet, but is it possible that the off by one error you changed could be the sole reason for sligh physics change?

**Comment by def-** [comment]
Coach is working on making this part of the map easier so it can be finished again.
